File: src/main/java-reference/de/visustruct/view/PaletteButtonStyle.java
package de.visustruct.view;

import java.awt.Color;
import java.awt.Component;

import javax.swing.JButton;
import javax.swing.SwingUtilities;
import javax.swing.UIManager;

import com.formdev.flatlaf.FlatClientProperties;

/**
 * Gleiche Darstellung aller Schaltflächen in der linken Palette (Farbe, Form, kein „Default“-Button).
 */
public final class PaletteButtonStyle {

	private static final int MIN_HEIGHT = 32;

	private PaletteButtonStyle() {
	}

	public static void apply(JButton b) {
		b.putClientProperty(FlatClientProperties.BUTTON_TYPE, FlatClientProperties.BUTTON_TYPE_ROUND_RECT);
		b.putClientProperty(FlatClientProperties.MINIMUM_HEIGHT, Integer.valueOf(MIN_HEIGHT));
		b.setDefaultCapable(false);
		b.setRolloverEnabled(true);
		b.setOpaque(true);
		b.setContentAreaFilled(true);
		b.setBorderPainted(true);

		Color bg = UIManager.getColor("Button.background");
		Color fg = UIManager.getColor("Button.foreground");
		if (bg != null) {
			b.setBackground(bg);
		}
		if (fg != null) {
			b.setForeground(fg);
		}
	}

	/**
	 * Nach Drag&amp;Drop bleibt bei {@link JButton} oft {@code pressed}/{@code armed} hängen — FlatLaf
	 * zeichnet dann dauerhaft grau. Aufruf nach Beendigung des Drags oder in {@code mouseReleased}.
	 */
	public static void clearPressedArmedState(Component c) {
		if (!(c instanceof JButton)) {
			return;
		}
		JButton b = (JButton) c;
		Runnable reset = () -> {
			b.getModel().setPressed(false);
			b.getModel().setArmed(false);
			b.repaint();
		};
		if (SwingUtilities.isEventDispatchThread()) {
			reset.run();
		} else {
			SwingUtilities.invokeLater(reset);
		}
	}
}

File: src/main/kotlin/de/visustruct/view/PaletteButtonStyle.kt
package de.visustruct.view

import com.formdev.flatlaf.FlatClientProperties
import java.awt.Color
import java.awt.Component
import javax.swing.JButton
import javax.swing.SwingUtilities
import javax.swing.UIManager

/** Gleiche Darstellung aller Schaltflächen in der linken Palette (Farbe, Form, kein „Default“-Button). */
object PaletteButtonStyle {

    private const val MIN_HEIGHT = 32

    @JvmStatic
    fun apply(b: JButton) {
        b.putClientProperty(FlatClientProperties.BUTTON_TYPE, FlatClientProperties.BUTTON_TYPE_ROUND_RECT)
        b.putClientProperty(FlatClientProperties.MINIMUM_HEIGHT, MIN_HEIGHT)
        b.isDefaultCapable = false
        b.isRolloverEnabled = true
        b.isOpaque = true
        b.isContentAreaFilled = true
        b.isBorderPainted = true
        val bg = UIManager.getColor("Button.background")
        val fg = UIManager.getColor("Button.foreground")
        if (bg != null) {
            b.background = bg
        }
        if (fg != null) {
            b.foreground = fg
        }
    }

    @JvmStatic
    fun clearPressedArmedState(c: Component?) {
        if (c !is JButton) {
            return
        }
        val reset = Runnable {
            c.model.isPressed = false
            c.model.isArmed = false
            c.repaint()
        }
        if (SwingUtilities.isEventDispatchThread()) {
            reset.run()
        } else {
            SwingUtilities.invokeLater(reset)
        }
    }
}
